### ISSUE TITLE
Windows 10: renaming of an empty folder triggers only once

### ISSUE THREAD

**Issue by AndyOGo** [paulmillr/chokidar#593]
If I rename a given folder from `foo` to `bar`, then to `baz` and back to `bar`.
Only the first `bar` rename is emitted, subsequent renamings to an already used name are not emitted any more, like jQuery's `once` event API.

Observed with chokidar `1.6.1`

**Comment by es128** [comment]
Please show what's being emitted out of the `raw` event while you're doing this.

**Comment by es128** [comment]
Also, if you want an immediate fix try the `usePolling: true` option. It's likely to resolve the issue for you, but comes with some performance tradeoffs.

**Comment by AndyOGo** [comment]
@es128 
Thanks for your quick reply.
I'm sorry for my late answer, I'm very busy atm and I wanted to supply a full failing test.

I'm the author of `node-sync-glob` which uses awesome `chokidar` under the hood.

So I prepared my lib and created a [dedicated test case regarding of deleting empty sub directories](https://github.com/AndyOGo/node-sync-glob/pull/46).

**TLDR:**

_Build Matrices_
- OS: MacOS, Linux, Windows (x86, x64)
- Node Version: Stable, 6, 5, 4

| OS | Result | CI Build |
| --- | --- | --- |
| MacOS | ✔️ | https://travis-ci.org/AndyOGo/node-sync-glob/jobs/220915576 |
| Linux | :x: | https://travis-ci.org/AndyOGo/node-sync-glob/jobs/220915572 |
| Windows | :x: | https://ci.appveyor.com/project/AndyOGo/node-sync-glob/build/1.0.107/job/wm7ualjc5nmwnlwo |

---

# MacOS Logs

Start at https://travis-ci.org/AndyOGo/node-sync-glob/jobs/220915576#L952

```sh
$ node --version
v7.8.0

$ npm --version
4.2.0

$ npm run test

> sync-glob@1.3.7 test /Users/travis/build/AndyOGo/node-sync-glob
> jest --runInBand --verbose --no-cache --config ./.jestrc.json

 PASS  test/lib/is-glob.spec.js
  lib/is-glob
    truthy
      ✓ should match asterisk wildcard (4ms)
      ✓ should match globstar wildcard (1ms)
      ✓ should match question mark wildcard (2ms)
      ✓ should match braced sections
      ✓ should match range wildcards (2ms)
      ✓ should match extended globs (1ms)

    falsy
      ✓ should not match escaped asterisks
      ✓ should not match escaped globstar (1ms)
      ✓ should not match escaped question mark
      ✓ should not match escaped braced sections (1ms)
      ✓ should not match escaped range wildcards
      ✓ should not match escaped extended globs (1ms)
      ✓ should not match non-globs

 PASS  test/copy.spec.js
  node-sync-glob copy
    ✓ should copy a file (49ms)
    ✓ should copy an array of files (22ms)
    ✓ should copy a directory (without contents) (85ms)
    ✓ should copy globs (53ms)
    ✓ should copy globstar (165ms)
    ✓ should copy empty sub directories (83ms)
    ○ skipped 1 test

console.log test/sync.spec.js:99
    EXISTS: tmp/mock/bar -> true

console.log src/index.js:102
    sources: tmp/mock/**/* -> tmp/mock/**/*

console.log src/index.js:103
    target: tmp/sync -> tmp/sync

console.log src/index.js:104
    globed files: 
    	tmp/mock/@org
    	tmp/mock/@org/a.txt
    	tmp/mock/@org/b.txt
    	tmp/mock/@org/d.txt
    	tmp/mock/a.txt
    	tmp/mock/b.txt
    	tmp/mock/bar
    	tmp/mock/bar/c.txt
    	tmp/mock/bar/empty
    	tmp/mock/foo
    	tmp/mock/foo/b.txt
    	tmp/mock/foo/d.txt
    	tmp/mock/transform.js

 PASS  test/sync.spec.js (10.163s)
  node-sync-glob watch
    ✓ should sync a file (2358ms)
    ✓ should sync an array of files (2510ms)
    ✓ should sync a directory (2439ms)
    ✓ should sync globstar (2420ms)
    ✓ should sync empty sub directory deletion (207ms)

console.log src/index.js:242
    RAW: deleted -> /Users/travis/build/AndyOGo/node-sync-glob/tmp/mock/bar/empty 	
    { path: '/Users/travis/build/AndyOGo/node-sync-glob/tmp/mock/bar/empty',
      event: 'deleted',
      type: 'directory',
      changes: { inode: false, finder: false, access: false, xattrs: false },
      flags: 131840 }

console.log src/index.js:190
    ALL: unlinkDir -> tmp/mock/bar/empty 

 PASS  test/lib/sources-bases.spec.js
  lib/sources-bases
    ✓ should resolve the base path of globs (8ms)
    ✓ should resolve the base path of files (1ms)
    ✓ should leave directories unchanged (1ms)
    ✓ should ignore exclude patterns
    ✓ should list multiple distinct base paths
    ✓ should list common base baths no more than once (1ms)

 PASS  test/lib/resolve-target.spec.js
  lib/resolve-target
    ✓ should resolve targets (6ms)

 PASS  test/transform.spec.js
  node-sync-glob transform
    ✓ should transform a file (65ms)

console.log test/mock/transform.js:3
    transform tmp/trans/b.txt

 PASS  test/lib/trim-quotes.spec.js
  lib/trim-quotes
    ✓ should trim single quotes (2ms)
    ✓ should trim double quotes
    ✓ should not remove quotes within text (1ms)

Test Suites: 7 passed, 7 total
Tests:       1 skipped, 35 passed, 36 total
Snapshots:   0 total
Time:        14.348s

Ran all test suites.
```
# Linux Logs

```sh
$ node --version
v7.8.0

$ npm --version
4.2.0

$ npm run test

> sync-glob@1.3.7 test /home/travis/build/AndyOGo/node-sync-glob
> jest --runInBand --verbose --no-cache --config ./.jestrc.json

 PASS  test/lib/is-glob.spec.js
  lib/is-glob
    truthy
      ✓ should match asterisk wildcard (4ms)
      ✓ should match globstar wildcard (1ms)
      ✓ should match question mark wildcard (2ms)
      ✓ should match braced sections
      ✓ should match range wildcards (2ms)
      ✓ should match extended globs (1ms)

    falsy
      ✓ should not match escaped asterisks (1ms)
      ✓ should not match escaped globstar
      ✓ should not match escaped question mark (1ms)
      ✓ should not match escaped braced sections
      ✓ should not match escaped range wildcards (1ms)
      ✓ should not match escaped extended globs
      ✓ should not match non-globs (1ms)

 PASS  test/copy.spec.js
  node-sync-glob copy
    ✓ should copy a file (32ms)
    ✓ should copy an array of files (9ms)
    ✓ should copy a directory (without contents) (26ms)
    ✓ should copy globs (16ms)
    ✓ should copy globstar (19ms)
    ✓ should copy empty sub directories (20ms)
    ○ skipped 1 test

console.log test/sync.spec.js:99
    EXISTS: tmp/mock/bar -> true

console.log src/index.js:102
    sources: tmp/mock/**/* -> tmp/mock/**/*

console.log src/index.js:103
    target: tmp/sync -> tmp/sync

console.log src/index.js:104
    globed files: 
    	tmp/mock/@org
    	tmp/mock/@org/a.txt
    	tmp/mock/@org/b.txt
    	tmp/mock/@org/d.txt
    	tmp/mock/a.txt
    	tmp/mock/b.txt
    	tmp/mock/bar
    	tmp/mock/bar/c.txt
    	tmp/mock/bar/empty
    	tmp/mock/foo
    	tmp/mock/foo/b.txt
    	tmp/mock/foo/d.txt
    	tmp/mock/transform.js

console.log src/index.js:242
    RAW: rename -> empty 	
    { watchedPath: 'tmp/mock/bar/empty' }

console.log src/index.js:242
    RAW: rename -> empty 	
    { watchedPath: 'tmp/mock/bar/empty' }

console.log src/index.js:242
    RAW: rename -> empty 	
    { watchedPath: 'tmp/mock/bar' }

 FAIL  test/sync.spec.js (13.696s)
  ● node-sync-glob watch › should sync empty sub directory deletion
    Timeout - Async callback was not invoked within timeout specified by jasmine.DEFAULT_TIMEOUT_INTERVAL.
      at ontimeout (timers.js:386:14)
      at tryOnTimeout (timers.js:250:5)
      at Timer.listOnTimeout (timers.js:214:5)

  node-sync-glob watch
    ✓ should sync a file (2135ms)
    ✓ should sync an array of files (2125ms)
    ✓ should sync a directory (2153ms)
    ✓ should sync globstar (2149ms)
    ✕ should sync empty sub directory deletion (5008ms)

console.log src/index.js:242
    RAW: change -> a.txt 	
    { watchedPath: 'tmp/mock/@org/a.txt' }

console.log src/index.js:242
    RAW: rename -> a.txt 	
    { watchedPath: 'tmp/mock/@org/a.txt' }

console.log src/index.js:242
    RAW: rename -> a.txt 	
    { watchedPath: 'tmp/mock/@org/a.txt' }

console.log src/index.js:242
    RAW: rename -> a.txt 	
    { watchedPath: 'tmp/mock/@org' }

console.log src/index.js:242
    RAW: change -> b.txt 	
    { watchedPath: 'tmp/mock/@org/b.txt' }

console.log src/index.js:242
    RAW: rename -> b.txt 	
    { watchedPath: 'tmp/mock/@org/b.txt' }

console.log src/index.js:242
    RAW: rename -> b.txt 	
    { watchedPath: 'tmp/mock/@org/b.txt' }

console.log src/index.js:242
    RAW: rename -> b.txt 	
    { watchedPath: 'tmp/mock/@org' }

console.log src/index.js:242
    RAW: change -> d.txt 	
    { watchedPath: 'tmp/mock/@org/d.txt' }

console.log src/index.js:242
    RAW: rename -> d.txt 	
    { watchedPath: 'tmp/mock/@org/d.txt' }

console.log src/index.js:242
    RAW: rename -> d.txt 	
    { watchedPath: 'tmp/mock/@org/d.txt' }

console.log src/index.js:242
    RAW: rename -> d.txt 	
    { watchedPath: 'tmp/mock/@org' }

console.log src/index.js:242
    RAW: rename -> @org 	
    { watchedPath: 'tmp/mock/@org' }

console.log src/index.js:242
    RAW: rename -> @org 	
    { watchedPath: 'tmp/mock/@org' }

console.log src/index.js:242
    RAW: rename -> @org 	
    { watchedPath: 'tmp/mock' }

console.log src/index.js:242
    RAW: change -> a.txt 	
    { watchedPath: 'tmp/mock/a.txt' }

console.log src/index.js:242
    RAW: rename -> a.txt 	
    { watchedPath: 'tmp/mock/a.txt' }

console.log src/index.js:242
    RAW: rename -> a.txt 	
    { watchedPath: 'tmp/mock/a.txt' }

console.log src/index.js:242
    RAW: rename -> a.txt 	
    { watchedPath: 'tmp/mock' }

console.log src/index.js:242
    RAW: change -> b.txt 	
    { watchedPath: 'tmp/mock/b.txt' }

console.log src/index.js:242
    RAW: rename -> b.txt 	
    { watchedPath: 'tmp/mock/b.txt' }

console.log src/index.js:242
    RAW: rename -> b.txt 	
    { watchedPath: 'tmp/mock/b.txt' }

console.log src/index.js:242
    RAW: rename -> b.txt 	
    { watchedPath: 'tmp/mock' }

console.log src/index.js:242
    RAW: change -> c.txt 	
    { watchedPath: 'tmp/mock/bar/c.txt' }

console.log src/index.js:242
    RAW: rename -> c.txt 	
    { watchedPath: 'tmp/mock/bar/c.txt' }

console.log src/index.js:242
    RAW: rename -> c.txt 	
    { watchedPath: 'tmp/mock/bar/c.txt' }

console.log src/index.js:242
    RAW: rename -> c.txt 	
    { watchedPath: 'tmp/mock/bar' }

console.log src/index.js:242
    RAW: rename -> bar 	
    { watchedPath: 'tmp/mock/bar' }

console.log src/index.js:242
    RAW: rename -> bar 	
    { watchedPath: 'tmp/mock/bar' }

console.log src/index.js:242
    RAW: rename -> bar 	
    { watchedPath: 'tmp/mock' }

console.log src/index.js:242
    RAW: change -> b.txt 	
    { watchedPath: 'tmp/mock/foo/b.txt' }

console.log src/index.js:242
    RAW: rename -> b.txt 	
    { watchedPath: 'tmp/mock/foo/b.txt' }

console.log src/index.js:242
    RAW: rename -> b.txt 	
    { watchedPath: 'tmp/mock/foo/b.txt' }

console.log src/index.js:242
    RAW: rename -> b.txt 	
    { watchedPath: 'tmp/mock/foo' }

console.log src/index.js:242
    RAW: change -> d.txt 	
    { watchedPath: 'tmp/mock/foo/d.txt' }

console.log src/index.js:242
    RAW: rename -> d.txt 	
    { watchedPath: 'tmp/mock/foo/d.txt' }

console.log src/index.js:242
    RAW: rename -> d.txt 	
    { watchedPath: 'tmp/mock/foo/d.txt' }

console.log src/index.js:242
    RAW: rename -> d.txt 	
    { watchedPath: 'tmp/mock/foo' }

console.log src/index.js:242
    RAW: rename -> foo 	
    { watchedPath: 'tmp/mock/foo' }

console.log src/index.js:242
    RAW: rename -> foo 	
    { watchedPath: 'tmp/mock/foo' }

console.log src/index.js:242
    RAW: rename -> foo 	
    { watchedPath: 'tmp/mock' }

console.log src/index.js:242
    RAW: change -> transform.js 	
    { watchedPath: 'tmp/mock/transform.js' }

console.log src/index.js:242
    RAW: rename -> transform.js 	
    { watchedPath: 'tmp/mock/transform.js' }

console.log src/index.js:242
    RAW: rename -> transform.js 	
    { watchedPath: 'tmp/mock/transform.js' }

console.log src/index.js:242
    RAW: rename -> transform.js 	
    { watchedPath: 'tmp/mock' }

console.log src/index.js:242
    RAW: rename -> mock 	
    { watchedPath: 'tmp/mock' }

console.log src/index.js:242
    RAW: rename -> mock 	
    { watchedPath: 'tmp/mock' }

 PASS  test/lib/sources-bases.spec.js
  lib/sources-bases
    ✓ should resolve the base path of globs (8ms)
    ✓ should resolve the base path of files (1ms)
    ✓ should leave directories unchanged (1ms)
    ✓ should ignore exclude patterns
    ✓ should list multiple distinct base paths (1ms)
    ✓ should list common base baths no more than once (1ms)

 PASS  test/lib/resolve-target.spec.js
  lib/resolve-target
    ✓ should resolve targets (4ms)

console.log src/index.js:190
    ALL: unlink -> tmp/mock/@org/a.txt 

console.log src/index.js:190
    ALL: unlink -> tmp/mock/@org/b.txt 

console.log src/index.js:190
    ALL: unlink -> tmp/mock/@org/d.txt 

console.log src/index.js:190
    ALL: unlink -> tmp/mock/a.txt 

console.log src/index.js:190
    ALL: unlink -> tmp/mock/b.txt 

console.log src/index.js:190
    ALL: unlink -> tmp/mock/bar/c.txt 

console.log src/index.js:190
    ALL: unlink -> tmp/mock/foo/b.txt 

console.log src/index.js:190
    ALL: unlink -> tmp/mock/foo/d.txt 

console.log src/index.js:190
    ALL: unlink -> tmp/mock/transform.js 

console.log src/index.js:190
    ALL: unlinkDir -> /home/travis/build/AndyOGo/node-sync-glob/tmp/mock/@org 

console.log src/index.js:190
    ALL: unlinkDir -> /home/travis/build/AndyOGo/node-sync-glob/tmp/mock/foo 

console.error test/helpers.js:77
    error -> Error: expect(received).toBe(expected)

    Expected value to be (using ===):
      true
    Received:
      false

console.log test/helpers.js:80
    Error: expect(received).toBe(expected)

    Expected value to be (using ===):
      true
    Received:
      false
        at /home/travis/build/AndyOGo/node-sync-glob/test/sync.spec.js:112:89
        at /home/travis/build/AndyOGo/node-sync-glob/test/helpers.js:100:7
        at /home/travis/build/AndyOGo/node-sync-glob/src/index.js:222:9
        at tryCatcher (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/util.js:16:23)
        at Promise._settlePromiseFromHandler (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/promise.js:512:31)
        at Promise._settlePromise (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/promise.js:569:18)
        at Promise._settlePromise0 (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/promise.js:614:10)
        at Promise._settlePromises (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/promise.js:693:18)
        at Promise._fulfill (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/promise.js:638:18)
        at /home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/nodeback.js:42:21
        at CB (/home/travis/build/AndyOGo/node-sync-glob/node_modules/fs-extra/lib/remove/rimraf.js:57:5)
        at options.lstat (/home/travis/build/AndyOGo/node-sync-glob/node_modules/fs-extra/lib/remove/rimraf.js:81:14)
        at /home/travis/build/AndyOGo/node-sync-glob/node_modules/graceful-fs/polyfills.js:284:29
        at /home/travis/build/AndyOGo/node-sync-glob/node_modules/graceful-fs/polyfills.js:284:29
        at FSReqWrap.oncomplete (fs.js:114:15)
console.warn node_modules/bluebird/js/release/debuggability.js:870
    Unhandled rejection TypeError: Cannot read property 'addExpectationResult' of undefined
        at Env.fail (/home/travis/build/AndyOGo/node-sync-glob/node_modules/jest-jasmine2/vendor/jasmine-2.5.2.js:1021:24)
        at fail (/home/travis/build/AndyOGo/node-sync-glob/node_modules/jest-jasmine2/vendor/jasmine-2.5.2.js:3636:23)
        at /home/travis/build/AndyOGo/node-sync-glob/test/sync.spec.js:104:9
        at /home/travis/build/AndyOGo/node-sync-glob/test/helpers.js:84:9
        at notifyError (/home/travis/build/AndyOGo/node-sync-glob/src/index.js:64:48)
        at tryCatcher (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/util.js:16:23)
        at Promise._settlePromiseFromHandler (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/promise.js:512:31)
        at Promise._settlePromise (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/promise.js:569:18)
        at Promise._settlePromise0 (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/promise.js:614:10)
        at Promise._settlePromises (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/promise.js:689:18)
        at Async.Object.<anonymous>.Async._drainQueue (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/async.js:133:16)
        at Async.Object.<anonymous>.Async._drainQueues (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/async.js:143:10)
        at Immediate.Async.drainQueues (/home/travis/build/AndyOGo/node-sync-glob/node_modules/bluebird/js/release/async.js:17:14)
        at runCallback (timers.js:672:20)
        at tryOnImmediate (timers.js:645:5)
        at processImmediate [as _immediateCallback] (timers.js:617:5)

 PASS  test/transform.spec.js
  node-sync-glob transform
    ✓ should transform a file (23ms)

console.log test/mock/transform.js:3
    transform tmp/trans/b.txt

 PASS  test/lib/trim-quotes.spec.js
  lib/trim-quotes
    ✓ should trim single quotes
    ✓ should trim double quotes
    ✓ should not remove quotes within text (1ms)

Test Suites: 1 failed, 6 passed, 7 total
Tests:       1 failed, 1 skipped, 34 passed, 36 total
Snapshots:   0 total
Time:        16.417s

Ran all test suites.
```
